File: k9mail-library/src/main/java/com/fsck/k9/mail/transport/smtp/NegativeSmtpReplyException.java
package com.fsck.k9.mail.transport.smtp;


import com.fsck.k9.mail.MessagingException;


/**
 * Exception that is thrown when the server sends a negative reply (reply codes 4xx or 5xx).
 */
class NegativeSmtpReplyException extends MessagingException {
    private static final long serialVersionUID = 8696043577357897135L;

    private final int mReplyCode;
    private final String mReplyText;

    public NegativeSmtpReplyException(int replyCode, String replyText) {
        super((replyText != null && !replyText.isEmpty()) ?
                replyText : ("Negative SMTP reply: " + replyCode),
                isPermanentSmtpError(replyCode));
        mReplyCode = replyCode;
        mReplyText = replyText;
    }

    private static boolean isPermanentSmtpError(int replyCode) {
        return replyCode >= 500 && replyCode <= 599;
    }

    public int getReplyCode() {
        return mReplyCode;
    }

    public String getReplyText() {
        return mReplyText;
    }
}

File: mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/NegativeSmtpReplyException.kt
package com.fsck.k9.mail.transport.smtp

import net.thunderbird.core.common.exception.MessagingException

/**
 * Exception that is thrown when the server sends a negative reply (reply codes 4xx or 5xx).
 */
open class NegativeSmtpReplyException(
    val replyCode: Int,
    val replyText: String,
    val enhancedStatusCode: EnhancedStatusCode? = null,
) : MessagingException(
    buildErrorMessage(replyCode, replyText),
    isPermanentSmtpError(replyCode),
)

private fun buildErrorMessage(replyCode: Int, replyText: String): String {
    return replyText.ifEmpty { "Negative SMTP reply: $replyCode" }
}

private fun isPermanentSmtpError(replyCode: Int): Boolean {
    return replyCode in 500..599
}
